datafusion/src/optimizer/timestamp_evaluation.rs

```diff
@@ -0,0 +1,177 @@
+            }
+            _ => {
+                let expr = plan.expressions();
+
+                // apply the optimization to all inputs of the plan
+                let inputs = plan.inputs();
+                let new_inputs = inputs
+                    .iter()
+                    .map(|plan| self.optimize_with_datetime(*plan, date_time))
+                    .collect::<Result<Vec<_>>>()?;
+
+                utils::from_plan(plan, &expr, &new_inputs)
+            }
+        }
+    }
+}
+
+impl OptimizerRule for TimestampEvaluation {
+    fn optimize(
+        &self,
+        plan: &LogicalPlan,
+        props: &ExecutionProps,
+    ) -> Result<LogicalPlan> {
+        self.optimize_with_datetime(plan, &props.query_execution_start_time.unwrap())
+    }
+
+    fn name(&self) -> &str {
+        "timestamp_evaluation"
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::logical_plan::LogicalPlanBuilder;
+    use crate::test::*;
+
+    fn get_optimized_plan_formatted(plan: &LogicalPlan) -> String {
+        let rule = TimestampEvaluation::new();
+        let execution_props = ExecutionProps {
+            query_execution_start_time: Some(chrono::Utc::now()),
+        };
+
+        let optimized_plan = rule
+            .optimize(plan, &execution_props)
+            .expect("failed to optimize plan");
+        return format!("{:?}", optimized_plan);
+    }
+
+    #[test]
+    fn single_now() {
+        let table_scan = test_table_scan().unwrap();
+        let proj = vec![Expr::ScalarFunction {
+            args: vec![],
+            fun: BuiltinScalarFunction::Now,
+        }];
+        let plan = LogicalPlanBuilder::from(&table_scan)
+            .project(proj)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        let expected = "Projection: TimestampNanosecond(";
+        assert!(get_optimized_plan_formatted(&plan).starts_with(expected));
+    }
+
+    #[test]
+    fn double_now() {
+        let table_scan = test_table_scan().unwrap();
+        let proj = vec![
+            Expr::ScalarFunction {
+                args: vec![],
+                fun: BuiltinScalarFunction::Now,
+            },
+            Expr::Alias(
+                Box::new(Expr::ScalarFunction {
+                    args: vec![],
+                    fun: BuiltinScalarFunction::Now,
+                }),
+                "t2".to_string(),
+            ),
+        ];
+        let plan = LogicalPlanBuilder::from(&table_scan)
+            .project(proj)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        let actual = get_optimized_plan_formatted(&plan);
+        println!("output is {}", &actual);
+        let expected_start = "Projection: TimestampNanosecond(";
+        assert!(actual.starts_with(expected_start));
+
+        let expected_end = ") AS t2\
+             \n  TableScan: test projection=None";
+        assert!(actual.ends_with(expected_end));
```

Comment:
It would probably be good here to ensure the same timestamp value was produced in place `now()` was replaced.

Since you can specify what timestamp value goes in perhaps you could hard code it in the test (so you could compare the plan with a known constant string). Perhaps something like:

```
        let start_time  = Utc.ymd(2018, 7, 1).and_hms(6, 0, 0); // 2018-Jul-01 06:00

        let execution_props = ExecutionProps {
            query_execution_start_time: Some(start_time),
        };
```

Reply:
Great suggestion. How did i miss this? 🤕

Reply:
Done ✌️